File: 1220-count-vowels-permutation/1220-count-vowels-permutation.cpp
class Solution {
public:
    int countVowelPermutation(int n) {
     vector<vector<long long>>dp(n,vector<long long>(5,1));
        int mod=1e9+7;
        for(int i=n-2;i>=0;i--){
            dp[i][0]=(dp[i+1][1])%mod;
            dp[i][1]=(dp[i+1][0]%mod+dp[i+1][2]%mod)%mod;
            dp[i][2]=(dp[i+1][0]%mod+dp[i+1][1]%mod+dp[i+1][3]%mod+dp[i+1][4]%mod)%mod;
            dp[i][3]=(dp[i+1][2]%mod+dp[i+1][4]%mod)%mod;
            dp[i][4]=(dp[i+1][0])%mod;
        }
        return (dp[0][0]%mod+dp[0][1]%mod+dp[0][2]%mod+dp[0][3]%mod+dp[0][4]%mod)%mod;
    }
};

File: 1220-count-vowels-permutation/1220-count-vowels-permutation.py
class Solution:
    def countVowelPermutation(self, n: int) -> int:
        M = 10**9+7
        N = 5
        dp = [1]*N
        G = {'a':['e','i','u'],'e':['a','i'],'i':['e','o'],'o':['i'],'u':['i','o']}
        d = {'a':0,'e':1,'i':2,'o':3,'u':4}
        for i in range(1,n):
            new_dp = [0]*N
            for v in G:
                for w in G[v]:
                    new_dp[d[v]] = (new_dp[d[v]]+dp[d[w]])%M
            dp = new_dp
        return sum(dp)%M
        